mentoring/dashboard.py

```diff
@@ -0,0 +1,215 @@
+
+import json
+import logging
+
+from .dashboard_visual import DashboardVisualData
+from .mcq import MCQBlock
+from .sub_api import sub_api
+from webob import Response
+from xblock.core import XBlock
+from xblock.exceptions import XBlockNotFoundError
+from xblock.fields import Scope, Dict, List, String
+from xblock.fragment import Fragment
+from xblock.validation import ValidationMessage
+from xblockutils.helpers import child_isinstance
+from xblockutils.resources import ResourceLoader
+from xblockutils.studio_editable import StudioEditableXBlockMixin
+
+
+# Globals ###########################################################
+
+log = logging.getLogger(__name__)
+loader = ResourceLoader(__name__)
+
+
+def _(text):
+    """ A no-op to mark strings that we need to translate """
+    return text
+
+# Classes ###########################################################
+
+
+@XBlock.needs("i18n")
+class DashboardBlock(StudioEditableXBlockMixin, XBlock):
+    """
+    A block to summarize self-assessment results.
+    """
+    display_name = String(
+        display_name=_("Display Name"),
+        help=_("Display name for this module"),
+        scope=Scope.settings,
+        default=_('Self-Assessment Summary'),
+    )
+    mentoring_ids = List(
+        display_name=_("Mentoring Blocks"),
+        help=_(
+            "This should be an ordered list of the url_names of each mentoring block whose multiple choice question "
+            "values are to be shown on this dashboard. The list should be in JSON format. Example: {example_here}"
+        ).format(example_here='["2754b8afc03a439693b9887b6f1d9e36", "215028f7df3d4c68b14fb5fea4da7053"]'),
+        scope=Scope.settings,
+        default=""
+    )
+    color_codes = Dict(
+        display_name=_("Color Coding"),
+        help=_(
+            "You can optionally set a color for each expected value. Example: {example_here}"
+        ).format(example_here='{ "1": "red", "2": "yellow", 3: "green", 4: "lightskyblue" }'),
+        scope=Scope.content,
+    )
+    visual_rules = String(
+        display_name=_("Visual Representation"),
+        default="",
+        help=_("Optional: Enter the JSON configuration of the visual representation desired (Advanced)."),
+        scope=Scope.content,
+        multiline_editor=True,
+        resettable_editor=False,
+    )
+
+    editable_fields = ('display_name', 'mentoring_ids', 'color_codes', 'visual_rules')
+    css_path = 'public/css/dashboard.css'
+
+    def get_mentoring_blocks(self):
+        """
+        Generator returning the specified mentoring blocks, in order.
+
+        returns a list. Will insert None for every invalid mentoring block ID.
+        """
+        for url_name in self.mentoring_ids:
+            mentoring_id = self.scope_ids.usage_id.course_key.make_usage_key('problem-builder', url_name)
+            try:
+                yield self.runtime.get_block(mentoring_id)
+            except XBlockNotFoundError:
+                yield None
```

Comment:
This should cause some sort of validation warning when it happens. At the moment, adding a bogus block essentially gets ignored.